http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriberModels.scala

```diff
@@ -0,0 +1,215 @@
+        case "createCluster" => message.as[CreateCluster]
+        case other                => Left(DecodingFailure(s"invalid message type: ${other}", List.empty))
+      }
+    } yield value
+  }
+
+  implicit val stopUpdateMessageEncoder: Encoder[StopUpdateMessage] =
+    Encoder.forProduct3("messageType", "updatedMachineConfig", "clusterId")(
+      x => (x.messageType, x.updatedMachineConfig, x.clusterId)
+    )
+
+  implicit val clusterFollowupDetailsEncoder: Encoder[ClusterFollowupDetails] =
+    Encoder.forProduct2("clusterId", "clusterStatus")(x => (x.clusterId, x.clusterStatus))
+
+  implicit val clusterTransitionFinishedEncoder: Encoder[ClusterTransitionFinishedMessage] =
+    Encoder.forProduct2("messageType", "clusterFollowupDetails")(x => (x.messageType, x.clusterFollowupDetails))
+
+  //TODO: These google specific models json codec should be moved to a better place once Rob's GCE PR is in
+  implicit val clusterNameEncoder: Encoder[ClusterName] = Encoder.encodeString.contramap(_.value)
+  implicit val operationNameEncoder: Encoder[OperationName] = Encoder.encodeString.contramap(_.value)
+  implicit val ipEncoder: Encoder[IP] = Encoder.encodeString.contramap(_.value)
+  implicit val dataprocInfoEncoder: Encoder[DataprocInfo] =
+    Encoder.forProduct4("googleId", "operationName", "stagingBucket", "hostIp")(x => DataprocInfo.unapply(x).get)
+
+  implicit val clusterProjectAndNameEncoder: Encoder[ClusterProjectAndName] = Encoder.forProduct2(
+    "googleProject",
+    "clusterName"
+  )(x => ClusterProjectAndName.unapply(x).get)
+
+  implicit val createClusterEncoder: Encoder[CreateCluster] =
+    Encoder.forProduct18(
+      "messageType",
+      "id",
+      "clusterProjectAndName",
+      "serviceAccountInfo",
+      "dataprocInfo",
+      "auditInfo",
+      "properties",
+      "jupyterExtensionUri",
+      "jupyterUserScriptUri",
+      "jupyterStartUserScriptUri",
+      "userJupyterExtensionConfig",
+      "defaultClientId",
+      "clusterImages",
+      "scopes",
+      "welderEnabled",
+      "customClusterEnvironmentVariables",
+      "runtimeConfig",
+      "traceId")(x =>
+      (x.messageType,
+       x.id,
+       x.clusterProjectAndName,
+       x.serviceAccountInfo,
+       x.dataprocInfo,
+       x.auditInfo,
+       x.properties,
+       x.jupyterExtensionUri,
+       x.jupyterUserScriptUri,
+       x.jupyterStartUserScriptUri,
+       x.userJupyterExtensionConfig,
+       x.defaultClientId,
+       x.clusterImages,
+       x.scopes,
+       x.welderEnabled,
+       x.customClusterEnvironmentVariables,
+       x.runtimeConfig,
+       x.traceId
+      ))
+
+  implicit val leoPubsubMessageEncoder: Encoder[LeoPubsubMessage] = Encoder.instance { message =>
+    message match {
+      case m: StopUpdateMessage                => m.asJson
+      case m: ClusterTransitionFinishedMessage => m.asJson
+      case m: CreateCluster => m.asJson
+    }
+  }
+}
+
+sealed trait PubsubHandleMessageError extends NoStackTrace {
```

Comment:
same thought as with the codec, it may be worth moving the error handling object and trait to another file to keep the single responsibility principle for these files intact

Reply:
I'll move these back to the other file